src/Compilers/CSharp/Test/Semantic/Semantics/SimpleProgramsTests.cs

```diff
@@ -3034,5 +3103,268 @@ public void Yield_02()
+        {
+            var text = @"
+System.Console.WriteLine(0);
+
+using System;
+
+Console.WriteLine(1);
+";
+
+            var comp = CreateCompilation(text, options: TestOptions.DebugExe, parseOptions: DefaultParseOptions);
+
+            comp.VerifyDiagnostics(
+                // (4,1): error CS1529: A using clause must precede all other elements defined in the namespace except extern alias declarations
+                // using System;
+                Diagnostic(ErrorCode.ERR_UsingAfterElements, "using System;").WithLocation(4, 1),
+                // (6,1): error CS0103: The name 'Console' does not exist in the current context
+                // Console.WriteLine(1);
+                Diagnostic(ErrorCode.ERR_NameNotInContext, "Console").WithArguments("Console").WithLocation(6, 1)
+                );
+        }
+
+        [Fact]
+        public void OutOfOrder_10()
+        {
+            var text = @"
+System.Console.WriteLine(0);
+
+[module: MyAttribute]
+
+class MyAttribute : System.Attribute
+{}
+";
+
+            var comp = CreateCompilation(text, options: TestOptions.DebugExe, parseOptions: DefaultParseOptions);
+
+            comp.VerifyDiagnostics(
+                // (4,2): error CS1730: Assembly and module attributes must precede all other elements defined in a file except using clauses and extern alias declarations
+                // [module: MyAttribute]
+                Diagnostic(ErrorCode.ERR_GlobalAttributesNotFirst, "module").WithLocation(4, 2)
+                );
+        }
+
+        [Fact]
+        public void OutOfOrder_11()
+        {
+            var text = @"
+System.Console.WriteLine(0);
+
+extern alias A;
+";
+
+            var comp = CreateCompilation(text, options: TestOptions.DebugExe, parseOptions: DefaultParseOptions);
+
+            comp.VerifyDiagnostics(
+                // (4,1): error CS0439: An extern alias declaration must precede all other elements defined in the namespace
+                // extern alias A;
+                Diagnostic(ErrorCode.ERR_ExternAfterElements, "extern").WithLocation(4, 1)
+                );
+        }
+
+        [Fact]
+        public void OutOfOrder_12()
+        {
+            var text = @"
+extern alias A;
+using System;
+
+[module: MyAttribute]
+
+Console.WriteLine(1);
+
+class MyAttribute : System.Attribute
+{}
+";
+
+            var comp = CreateCompilation(text, options: TestOptions.DebugExe, parseOptions: DefaultParseOptions);
+
+            comp.VerifyDiagnostics(
+                // (2,1): hidden CS8020: Unused extern alias.
+                // extern alias A;
+                Diagnostic(ErrorCode.HDN_UnusedExternAlias, "extern alias A;").WithLocation(2, 1),
+                // (2,14): error CS0430: The extern alias 'A' was not specified in a /reference option
+                // extern alias A;
+                Diagnostic(ErrorCode.ERR_BadExternAlias, "A").WithArguments("A").WithLocation(2, 14)
+                );
+        }
```

Comment:
>} [](start = 8, length = 1)

Is there a test of a type (or namespace) followed by a local function?

Reply:
> Is there a test of a type (or namespace) followed by a local function?

No, those are no different than any other top-level statements - all of them a wrapped into GlobalStatementSyntax

---
In reply to: [386529251](https://github.com/dotnet/roslyn/pull/42008#discussion_r386529251) [](ancestors = 386529251)

Reply:
I think it would be useful to have such a test so it's clear that is that top-level methods are treated as top-level statements.

It would also make clear the error message is "Top-level statements must precede namespace and type declarations" which might be a little confusing when a member method is treated as a top-level statement because of a misplaced `}` before the method.

---
In reply to: [386553202](https://github.com/dotnet/roslyn/pull/42008#discussion_r386553202) [](ancestors = 386553202,386529251)

Reply:
> It would also make clear the error message is "Top-level statements must precede namespace and type declarations" which might be a little confusing when a member method is treated as a top-level statement because of a misplaced } before the method.

Is there any specific suggestion in this statement? Could you please elaborate?


---
In reply to: [386558178](https://github.com/dotnet/roslyn/pull/42008#discussion_r386558178) [](ancestors = 386558178,386553202,386529251)

Reply:
It's probably a common mistake to copy and paste methods unintentionally as top-level methods so perhaps the diagnostic message should mention methods explicitly. For instance: "Top-level statements and functions must precede namespace and type declarations."

---
In reply to: [386563032](https://github.com/dotnet/roslyn/pull/42008#discussion_r386563032) [](ancestors = 386563032,386558178,386553202,386529251)

Reply:
> For instance: "Top-level statements and functions must precede namespace and type declarations."

"Top-level functions" is a different feature. Local function are not them. I think, the message is clear enough, it points out that the statement is a top-level statement, so if that is not expected it is pretty clear what to look for (unexpected end of type declaration, etc.). I'll add a PROTOTYPE comment to follow-up if that would be necessary.

---
In reply to: [386566431](https://github.com/dotnet/roslyn/pull/42008#discussion_r386566431) [](ancestors = 386566431,386563032,386558178,386553202,386529251)